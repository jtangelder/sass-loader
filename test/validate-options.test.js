import Fiber from "fibers";

import {
  getCompiler,
  compile,
  getTestId,
  getImplementationByName,
} from "./helpers/index";

describe("validate options", () => {
  beforeEach(() => {
    // The `sass` (`Dart Sass`) package modify the `Function` prototype, but the `jest` lose a prototype
    Object.setPrototypeOf(Fiber, Function.prototype);
  });

  const tests = {
    implementation: {
      // eslint-disable-next-line global-require
      success: [require("sass"), require("node-sass")],
      failure: [true, "string"],
    },
    sassOptions: {
      success: [
        {},
        { indentWidth: 6 },
        () => {
          return { indentWidth: 6 };
        },
      ],
      failure: [true, "string"],
    },
    additionalData: {
      success: ["$color: red;", (content) => `$color: red;\n${content}`],
      failure: [true],
    },
    sourceMap: {
      success: [true, false],
      failure: ["string"],
    },
    webpackImporter: {
      success: [true, false],
      failure: ["string"],
    },
    unknown: {
      success: [],
      failure: [1, true, false, "test", /test/, [], {}, { foo: "bar" }],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === "object" && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      const testId = getTestId("language", "scss");
      const compiler = getCompiler(testId, {
        loader: {
          options: {
            implementation: getImplementationByName("dart-sass"),
            [key]: value,
          },
        },
      });
      let stats;

      try {
        stats = await compile(compiler);
      } finally {
        if (type === "success") {
          expect(stats.hasErrors()).toBe(false);
        } else if (type === "failure") {
          const {
            compilation: { errors },
          } = stats;

          expect(errors).toHaveLength(1);
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
