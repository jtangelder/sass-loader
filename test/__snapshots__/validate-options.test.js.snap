// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options 1`] = `
"Invalid options object. Sass Loader has been initialized using an options object that does not match the API schema.
 - options.implementation should be an object:
   object { … }
   -> The implementation of the sass to be used (https://github.com/webpack-contrib/sass-loader#implementation)."
`;

exports[`validate options 2`] = `
"Invalid options object. Sass Loader has been initialized using an options object that does not match the API schema.
 - options.sassOptions should be one of these:
   object { … } | function
   -> Options for \`node-sass\` or \`sass\` (\`Dart Sass\`) implementation. (https://github.com/webpack-contrib/sass-loader#implementation).
   Details:
    * options.sassOptions should be an object:
      object { … }
    * options.sassOptions should be an instance of function."
`;

exports[`validate options 3`] = `
"Invalid options object. Sass Loader has been initialized using an options object that does not match the API schema.
 - options.prependData should be one of these:
   string | function
   -> Prepends \`Sass\`/\`SCSS\` code before the actual entry file (https://github.com/webpack-contrib/sass-loader#prependdata).
   Details:
    * options.prependData should be a string.
    * options.prependData should be an instance of function."
`;

exports[`validate options 4`] = `
"Invalid options object. Sass Loader has been initialized using an options object that does not match the API schema.
 - options.webpackImporter should be a boolean.
   -> Enables/Disables default \`webpack\` importer (https://github.com/webpack-contrib/sass-loader#webpackimporter)."
`;

exports[`validate options 5`] = `
"Invalid options object. Sass Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { implementation?, sassOptions?, prependData?, sourceMap?, webpackImporter? }"
`;
